File: web/cgi/srchres.pl
#!/usr/bin/env perl
#######################################################################
#   This file is part of JMdictDB. 
#   Copyright (c) 2006,2007 Stuart McGraw 
# 
#   JMdictDB is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published 
#   by the Free Software Foundation; either version 2 of the License, 
#   or (at your option) any later version.
# 
#   JMdictDB is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
# 
#   You should have received a copy of the GNU General Public License
#   along with JMdictDB; if not, write to the Free Software Foundation,
#   51 Franklin Street, Fifth Floor, Boston, MA  02110#1301, USA
#######################################################################

@VERSION = (substr('$Revision$',11,-2), \
	    substr('$Date$',7,-11));

use strict; use warnings;
use Cwd; use CGI; use Encode 'decode_utf8'; use DBI; 
use Petal; use Petal::Utils; use Time::HiRes('time');

use lib ("../lib", "./lib", "../perl/lib");
use jmdict; use jmdicttal; use jmdictcgi;

$|=1;
binmode (STDOUT, ":utf8");

    main: {
	my ($dbh, $cgi, $tmpl, @s, @y, @t, $col, @kinf, @rinf, @fld, $svc, $svcstr,
	    @pos, @misc, @src, @stat, @freq, $nfval, $nfcmp, $gaval, $gacmp, @appr, 
	    $idval, $idtbl, $sql, $sql_args, $sql2, $rs, $i, $freq, @condlist,
	    $force_srchres);
	binmode (STDOUT, ":encoding(utf-8)");
	$cgi = new CGI;
	$svc = clean ($cgi->param ("svc"));
	$dbh = dbopen ($svc);  $::KW = Kwds ($dbh);

	$s[0]=$cgi->param("s1"); $y[0]=$cgi->param("y1"); $t[0]=decode_utf8($cgi->param("t1"));
	$s[1]=$cgi->param("s2"); $y[1]=$cgi->param("y2"); $t[1]=decode_utf8($cgi->param("t2"));
	$s[2]=$cgi->param("s3"); $y[2]=$cgi->param("y3"); $t[2]=decode_utf8($cgi->param("t3"));
	@pos=$cgi->param("pos");   @misc=$cgi->param("misc"); @fld=$cgi->param("fld");
	@rinf=$cgi->param("rinf"); @kinf=$cgi->param("kinf"); @freq=$cgi->param("freq");
	@src=$cgi->param("src");   @stat=$cgi->param("stat"); @appr=$cgi->param('appr'); 
	$nfval=$cgi->param("nfval"); $nfcmp=$cgi->param("nfcmp");
	$gaval=$cgi->param("gaval"); $gacmp=$cgi->param("gacmp");
	$idval=$cgi->param("idval"); $idtbl=$cgi->param("idtyp");
	$force_srchres = $cgi->param ("srchres"); # Force display of srchres page even if only one result.

	# The followng will convert substrings like '\u6a8e' into
	# a unicode character.  Sequences like this often occur
	# when pasting text between remote unix VNC sessions, and
	# a MS Windows application.
	1 while $t[0] =~ s/(\\u([0-9a-f]{4}))/chr(hex($2))/ei;	
	1 while $t[1] =~ s/(\\u([0-9a-f]{4}))/chr(hex($2))/ei;	
	1 while $t[2] =~ s/(\\u([0-9a-f]{4}))/chr(hex($2))/ei;	

	if ($idval) {	# Search for id number...
	    if ($idtbl ne "seqnum") { $col = "id"; }
	    else { $idtbl = "entr e";  $col = "seq"; }
	    push (@condlist, [$idtbl, sprintf ("e.%s=?", $col), [$idval]]);
	    if ($col eq "seq" and @src)  { 
		push (@condlist, ["entr e",getsel("e.src", \@src), []]); }
	    ($sql, $sql_args) = build_search_sql (\@condlist); }
	else {
	    for $i (0..2) {
		if ($t[$i]) { 
		    push (@condlist, str_match_clause ($s[$i],$y[$i],$t[$i],$i)); } }
	    if (@pos)  { push (@condlist, ["pos",   getsel("pos.kw",  \@pos), []]); }
	    if (@misc) { push (@condlist, ["misc",  getsel("misc.kw", \@misc),[]]); }
	    if (@fld)  { push (@condlist, ["fld",   getsel("fld.kw",  \@fld), []]); }
	    if (@kinf) { push (@condlist, ["kinf",  getsel("kinf.kw", \@kinf),[]]); }
	    if (@rinf) { push (@condlist, ["rinf",  getsel("rinf.kw", \@rinf),[]]); }
	    if (@src)  { push (@condlist, ["entr e",getsel("e.src",   \@src), []]); }
	    if (@stat) { push (@condlist, ["entr e",getsel("e.stat",  \@stat),[]]); }
	    if (@appr) { push (@condlist, ["entr e",getbool("e.unap",  \@appr),[]]); }
	    if (@freq) { push (@condlist, freq_srch_clause (\@freq, $nfval, $nfcmp, $gaval, $gacmp)); }
	    ($sql, $sql_args) = build_search_sql (\@condlist); }

	$::Debug->{'Search sql'} = $sql;  $::Debug->{'Search args'} = join(",", @$sql_args);
	$sql2 = sprintf ("SELECT q.* FROM esum q JOIN (%s) AS i ON i.id=q.id", $sql);
	my $start = time();
	eval { $rs = dbread ($dbh, $sql2, $sql_args); };
	$::Debug->{'Search time'} = time() - $start;
	if ($@) {
	    print "Content-type: text/html\n\n<html><head><meta http-equiv=\"Content-Type\" content=\"text/html; charset=utf-8\"/></head><body>";
	    print "<pre> $@ </pre>\n<pre>$sql2</pre>\n<pre>".join(", ", @$sql_args)."</pre></body></html>\n";
	    exit (1); }
	if (scalar (@$rs) == 1 && !$force_srchres) {
	    $svcstr = $svc ? "svc=$svc&" : "";
	    printf ("Location: entr.pl?${svcstr}e=%d\n\n", $rs->[0]{id}); }
	else {
	    print "Content-type: text/html\n\n";
	    $tmpl = new Petal (file=>find_in_inc("tal")."/tal/srchres.tal", 
			   decode_charset=>'utf-8', output=>'HTML' );
	    print $tmpl->process (results=>$rs, svc=>$svc, dbg=>$::Debug); }
	$dbh->disconnect; }

    sub str_match_clause { my ($srchin, $srchtyp, $srchtxt, $idx) = @_; 
	my ($x, $table, $alias, $whr, @args, $column);
	if ($srchin eq "auto") {
	    $x = jstr_classify ($srchtxt);
	    if ($x & $jmdict::KANJI) { $table = "kanj";  }
	    elsif ($x & $jmdict::KANA) { $table = "rdng"; }
	    else { $table = "gloss"; } }
	else { $table = $srchin; }
	$alias = " " . {kanj=>"j", rdng=>"r", gloss=>"g"}->{$table} . "$idx"; 
	$srchtyp = lc($srchtyp);

	  # The following generates implements case insensitive search
	  # for gloss searches, and non-"is" searches using the sql LIKE
	  # operator.  The case-insensitive part is a work-around for
	  # Postgresql's lack of support for standard SQL's COLLATION
	  # feature.  We can't use ILIKE for case-insensitive searches
	  # because it won't use an index and thus is very slow (~25s
	  # vs ~.2s with index on developer's machine.  So instead, we
	  # created two functional indexes on gloss.txt: "lower(txt)"
	  # and "lower(txt) varchar-pattern-ops".  The former will be
	  # used for "lower(xx)=..." searches and the latter for
	  # "lower(xx) LIKE ..." searches.  So when do a gloss search,
	  # we need to lowercase the search text, and generate a search
	  # clause in one of the above forms.  
	  #
	  # To-do: LIKE 'xxx%' dosn't use index unless the argument value 
	  # is embedded in the sql (which we don't currently do).  When
	  # the 'xxx%' is supplied as a separate argument, the query
	  # planner (runs when the sql is parsed) can't use index because
	  # it doesn't have access to the argument (which is only available
	  # when the query is executed) and doesn't know that it is not
	  # something like '%xxx'.

	if ($table eq "gloss") {
	    $srchtxt = lc ($srchtxt); 
	    $column = "lower($alias.txt)"; }
	else { $column = "$alias.txt"; }
	if ($srchtyp eq "is") { $whr = "$column=?"; }
	else { $whr = "$column LIKE(?)"; }

	  # Now generate the argument list appropriate for the
	  # search clause.

	if ($srchtyp eq "is") 		{ @args = ($srchtxt); } 
	elsif ($srchtyp eq "starts")	{ @args = ($srchtxt."%",); }
	elsif ($srchtyp eq "contains")	{ @args = ("%".$srchtxt."%"); }
	elsif ($srchtyp eq "ends")	{ @args = ("%".$srchtxt); }
	else { die "Unknown srchtyp value encountered in str_match_clause(): $srchtyp\n"; }

	return ["$table$alias",$whr,\@args]; }

    sub getsel { my ($fqcol, $itms) = @_;
	my $s = sprintf ("%s IN (%s)", $fqcol, join(",", map (int($_), @$itms)));
	return $s; }

    sub getbool { my ($fqcol, $itms) = @_;
	my $s = join (" OR ", map (int($_) ? $fqcol : "NOT $fqcol ", @$itms));
	return "($s)"; }

    sub freq_srch_clause { my ($freq, $nfval, $nfcmp, $gaval, $gacmp) = @_;
	# Create a pair of 3-tuples (build_search_sql() "conditions")
	# that build_search_sql() will use to create a sql statement 
	# that will incorporate the freq-of-use criteria defined by
	# our parameters:
	#
	# $freq -- List of string values of a freq option checkboxes, e.g. "ichi2".
	# $nfval -- String containing an "nf" number ("1" - "48").
	# $nfcmp -- String containing one of ">=", "=", "<=".
	# gaval -- String containing a gA number.
	# gacmp -- Same as nfcmp.

	my ($f, $domain, $value, %x, $k, $v, $kwid, @whr, $whr);

	# Freq items consist of a domain (such as "ichi" or "nf")
	# and a value (such as "1" or "35").
	# Process the checkboxes by creating a hash indexed by 
	# by domain and with each value a list of freq values.

	foreach $f (@$freq) {
	    # Split into text (domain) and numeric (value) parts.
	    ($domain, $value) = ($f =~ m/(^[A-Za-z_-]+)(\d*)$/);
	    # We will handle "nfxx" and "gAxxxx" later.
	    next if ($domain eq "nf" or $domain eq "ga");
	    # If this domain not in hash yet, add it.
	    if (!defined ($x{$domain})) { $x{$domain} = []; }
	    # Append this value to the list.
	    push (@{$x{$domain}}, $value); }

	# Now process each domain and it's list of values...

	while (($k,$v) = each (%x)) {
	    # Convert the domain string to a kwfreq table id number.
	    $kwid = $::KW->{FREQ}{$k}{id};

	    # The following assumes that the range of values are 
	    # limited to 1 and 2.

	    if (scalar(@$v)==2) { push (@whr, sprintf (
		# As an optimization, if there are 2 values, they must be 1 and 2, 
		# so no need to check value in query, just see if the domain exists.
		# FIXME: The above is false, there could be two "1" values.
		# FIXME: The above assumes only 1 and 2 are allowed.  Currently
		#   true but may change in future.
		"(freq.kw=%s)", $kwid)); }
	    elsif (scalar(@$v) == 1) { push (@whr, sprintf (
		# If there is only one value we need to look for kw with
		# that value.
		"(freq.kw=%s AND freq.value=%s)", $kwid, $v->[0])); }
	    elsif (scalar(@$v) > 2) { push (@whr, sprintf (
		# If there are more than 2 values then we look for them explicitly
		# using an IN() construct.
		"(freq.kw=%s AND freq.value IN (%s))", $k, join(",",@$v))); }
	    # A 0 or negative length list should be impossible.
	    else { die; } }

	# Handle the "nfxx" items specially here.

	if (grep ($_ eq "nf", @$freq) and $nfval) {
	    # Convert the domain string to a kwfreq table id number.
	    $kwid = $::KW->{FREQ}{nf}{id};
	    # Build list of "where" clause parts using the requested comparison and value.
	    push (@whr, sprintf (
		"(freq.kw=%s AND freq.value%s%s)", $kwid, $nfcmp, $nfval)); }

	# Handle the "gAxx" items specially here.

	if (grep ($_ eq "ga", @$freq) and $gaval) {
	    # Convert the domain string to a kwfreq table id number.
	    $kwid = $::KW->{FREQ}{gA}{id};
	    # Build list of "where" clause parts using the requested comparison and value.
	    push (@whr, sprintf (
		"(freq.kw=%s AND freq.value%s%s)", $kwid, $gacmp, $gaval)); }

	# Now, @whr is a list of all the various freq ewlated conditions that 
	# were  selected.  We change it into a clause by connecting them all 
	# with " OR".
	$whr = "(" . join(" OR ", @whr) . ")";

	# If there were no freq related conditions...
	return [] if (!$whr);

	# Return two triples suitable for use by build-search_sql().  That function
	# will build sql that effectivly "AND"s all the conditions (each specified 
	# in a triple) given to it.  Our freq conditions applies to two tables 
	# (rfreq and kfreq) and we want them OR'd not AND'd.  So we cheat and use a
	# strisk in front of table name to tell build_search_sql() to use left joins
	# rather than inner joins when refering to that condition's table.  This will
	# result in the inclusion in the result set of rfreq rows that match the
	# criteria, even if there are no matching kfreq rows (and visa versa). 
	# The where clause refers to both the rfreq and kfreq tables, so need only
	# be given in one constion triple rather than in each. 
	return (["freq",$whr,[]]); }

File: web/cgi/srchres.py
#!/usr/bin/env python
#######################################################################
#  This file is part of JMdictDB. 
#  Copyright (c) 2006-2010 Stuart McGraw 
# 
#  JMdictDB is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published 
#  by the Free Software Foundation; either version 2 of the License, 
#  or (at your option) any later version.
# 
#  JMdictDB is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
# 
#  You should have received a copy of the GNU General Public License
#  along with JMdictDB; if not, write to the Free Software Foundation,
#  51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
#######################################################################
from __future__ import print_function

__version__ = ('$Revision$'[11:-2],
	       '$Date$'[7:-11])

import sys, cgi, copy, time
sys.path.extend (['../lib','../../python/lib','../python/lib'])
import cgitbx; cgitbx.enable()
import jdb, jmcgi, serialize, jelparse

def main( args, opts ):
	errs = []; so = None; stats = {}
	try: form, svc, host, cur, sid, sess, parms, cfg = jmcgi.parseform()
	except StandardError as e: jmcgi.err_page ([unicode (e)])

	cfg_web = d2o (cfg['web'])
	cfg_srch = d2o (cfg['search'])
	fv = form.getfirst; fl = form.getlist
	dbg = fv ('d'); meth = fv ('meth')
	force_srchres = fv('srchres')  # Force display of srchres page even if only one result.
	sqlp = (fv ('sql') or '').decode ('utf-8')
	soj = (fv ('soj') or '').decode ('utf-8')
	pgoffset = int(fv('p1') or 0)
	pgtotal = int(fv('pt') or -1)
	entrs_per_page = min (max (int(fv('ps') or cfg_web.DEF_ENTRIES_PER_PAGE),
			   cfg_web.MIN_ENTRIES_PER_PAGE), cfg_web.MAX_ENTRIES_PER_PAGE)
	if not sqlp and not soj:
	    so = jmcgi.SearchItems()
	    so.idnum=fv('idval');  so.idtyp=fv('idtyp')
	    tl = []
	    for i in (1,2,3):
	        txt = (fv('t'+str(i)) or '').decode('utf-8')
	        if txt: tl.append (jmcgi.SearchItemsTexts (
				     srchtxt = txt, 
				     srchin  = fv('s'+str(i)),
				     srchtyp = fv('y'+str(i)) ))
	    if tl: so.txts = tl
	    so.pos   = fl('pos');   so.misc  = fl('misc');  
	    so.fld   = fl('fld');   so.dial  = fl('dial');
	    so.rinf  = fl('rinf');  so.kinf  = fl('kinf');  so.freq = fl('freq')
	    so.grp   = grpsparse (fv('grp'))
	    so.src   = fl('src');   so.stat  = fl('stat');  so.unap = fl('appr')
	    so.nfval = fv('nfval'); so.nfcmp = fv('nfcmp')
	    so.gaval = fv('gaval'); so.gacmp = fv('gacmp')
	      #FIXME? use selection boxes for dates?  Or a JS calendar control?
	    so.ts = dateparse (fv('ts0'), 0, errs), dateparse (fv('ts1'), 1, errs)
	    so.smtr = (fv('smtr') or '').decode('utf-8'), fv('smtrm')
	    so.mt = fv('mt')
	      # Pack up all the search criteria in a json string that will 
	      # be given to the srchres form, which will in turn give it back
	      # to us if the user want to display the "next page".  
	    soj = serialize.so2js (so)

	elif soj:
	      # 'soj' is a json string that encodes the so object (containing
	      # the search criteria) that were used in previous invocation
	      # of this script, which displayed the previous page.
	    so = serialize.js2so (soj)

	elif sqlp:
	      # 'sqlp' is a SQL statement string that allows an arbitrary search.
	      # Because it can also do other things such as delete the database,
	      # it should only be run as a user with read-only access to the
	      # database and it is the job of jmcgi.adv_srch_allowed() to check
	      # that.
	    if not jmcgi.adv_srch_allowed (cfg, sess):
		jmcgi.err_page (["'sql' parameter is disallowed."])
	    sql = sqlp.strip()
	    if sql.endswith (';'): sql = sql[:-1]
	    sql_args = []

	if so:
	    try: condlist = jmcgi.so2conds (so)
	    except ValueError as e:
		errs.append (unicode (e))
	      # FIXME: [IS-115] Following will prevent kanjidic entries from
	      #  appearing in results.  Obviously hardwiring id=4 is a hack.
	    else:
	        #condlist.append (('entr e', 'e.src!=4', []))
	        sql, sql_args = jdb.build_search_sql (condlist)

	if errs: jmcgi.err_page (errs)

	orderby = "ORDER BY __wrap__.kanj,__wrap__.rdng,__wrap__.seq,__wrap__.id"
        page = "OFFSET %s LIMIT %s" % (pgoffset, entrs_per_page)
	sql2 = "SELECT __wrap__.* FROM esum __wrap__ " \
		 "JOIN (%s) AS __user__ ON __user__.id=__wrap__.id %s %s" \
		  % (sql, orderby, page)
	stats['sql']=sql; stats['args']=sql_args; stats['orderby']=orderby
	if cfg_srch.MAX_QUERY_COST > 0:
	    try:
	        cost = jdb.get_query_cost (cur, sql2, sql_args);
	    except StandardError as e:
		jmcgi.err_page (["Database error (%s):<pre> %s </pre></body></html>" 
			   % (e.__class__.__name__, str(e))])
	    stats['cost']=cost;
	    if cost > cfg_srch.MAX_QUERY_COST: 
		jmcgi.err_page (
		       ["The search request you made will likely take too long to execute. "
			"Please use your browser's \"back\" button to return to the search "
			"page and add more criteria to restrict your search more narrowly. "
			"(The estimated cost was %.1f, max allowed is %d.)" 
			% (cost, cfg_srch.MAX_QUERY_COST)])
	t0 = time.time()
	try: rs = jdb.dbread (cur, sql2, sql_args)
	except Exception as e:		#FIXME, what exception value(s)?
	    jmcgi.err_page (["Database error (%s):<pre> %s </pre></body></html>" 
		       % (e.__class__.__name__, str(e))])
	stats['dbtime'] = time.time() - t0
	reccnt = len(rs)
	if pgtotal < 0:
	    if reccnt >= entrs_per_page:
	      # If there may be more than one page of entries (because
	      # 'reccnt' is greater than the page size, 'entrs_per_page',
	      # then run another query to get the actual number of entries.
	      # We only do this on the first page of results ('pgtotal' is
	      # less then 0) and subsequently pass the value between pages
	      # for performace reasons, even though the number of entries
	      # may change before the user gets to the last page.
	        sql3 = "SELECT COUNT(*) AS cnt FROM (%s) AS i " % sql
	        cntrec = jdb.dbread (cur, sql3, sql_args)
	        pgtotal = cntrec[0][0]	# Total number of entries.
	    else: pgtotal = reccnt
	if reccnt == 1 and pgtotal == 1 and not force_srchres:
	      # If there is only one entry, display it rather than a search
	      # results page.  'force_srchres' allows supressing this behavior
	      # for debugging.
	    svcstr = ("svc=%s&sid=%s&" % (svc,sid)) if svc else ''
	    print ("Location: entr.py?%se=%d\n" % (svcstr, rs[0].id))
	else:
	    if not meth: meth = 'get' if dbg else 'post'
	    jmcgi.gen_page ("tmpl/srchres.tal", macros='tmpl/macros.tal', 
			    results=rs, pt=pgtotal, p0=pgoffset, method=meth,
			    p1=pgoffset+reccnt, soj=soj, sql=sqlp, parms=parms,
			    svc=svc, host=host, sid=sid, session=sess, cfg=cfg,
			    stats=stats, output=sys.stdout, this_page='srchres.py')

def d2o (dict_):
	# Copy the key/value items in a dict to attributes on an 
	# object, converting numbers to ints when possible.
	# FIXME: What about floats, bools, datetimes, lists, ...?
	#  Should we consider JSON as an ini file format?
	o = jdb.Obj()
	for k,v in dict_.items():
	    try: v = int (v)
	    except (ValueError,TypeError): pass
	    setattr (o, k, v)
	return o

def grpsparse (grpsstr):
	if not grpsstr: return []
	return grpsstr.split()

def dateparse (dstr, upper, errs): 
	if not dstr: return None 
	dstr = dstr.strip();  dt = None
	if not dstr: return None 
	  # Add a time if it wasn't given.
	if len(dstr) < 11: dstr += " 23:59" if upper else " 00:00"
	  # Note: we use time.strptime() to parse because it returns 
	  # struct easily converted into a 9-tuple, which in turn is 
	  # easily JSONized, unlike a datetime.datetime object. 
	try: dt = time.strptime (dstr, "%Y/%m/%d %H:%M")
	except ValueError:
	    try: dt = time.strptime (dstr, "%Y-%m-%d %H:%M")
	    except ValueError: 
		errs.append ("Unable to parse date/time string '%s'." % cgi.escape(dstr))
	if dt: return time.mktime (dt)
	return None

if __name__ == '__main__': 
	args, opts = jmcgi.args()
	main (args, opts)
